sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java

```diff
@@ -279,28 +314,73 @@ public void processElement(ProcessContext processContext) throws Exception {
       while (!finished) {
         ScanResult<String> scanResult = jedis.scan(cursor, scanParams);
         List<String> keys = scanResult.getResult();
-
-        Pipeline pipeline = jedis.pipelined();
-        if (keys != null) {
-          for (String key : keys) {
-            pipeline.get(key);
-          }
-          List<Object> values = pipeline.syncAndReturnAll();
-          for (int i = 0; i < values.size(); i++) {
-            processContext.output(KV.of(keys.get(i), (String) values.get(i)));
-          }
+        for (String k : keys) {
+          processContext.output(k);
         }
-
         cursor = scanResult.getStringCursor();
         if ("0".equals(cursor)) {
           finished = true;
         }
       }
     }
+  }
+  /** A {@link DoFn} requesting Redis server to get key/value pairs. */
+  private static class ReadFn extends BaseReadFn<KV<String, String>> {
+    private int batchSize;
+    private List<String> bufferedKeys;
+    BoundedWindow window;
+    Instant lastMsg;
+
+    @StartBundle
+    public void startBundle(StartBundleContext context) {
+      bufferedKeys = new ArrayList<>();
+    }
 
-    @Teardown
-    public void teardown() {
-      jedis.close();
+    ReadFn(RedisConnectionConfiguration connectionConfiguration, int batchSize) {
+      super(connectionConfiguration);
+      this.batchSize = batchSize;
+    }
+
+    private int getBatchSize() {
+      return batchSize;
+    }
+
+    @ProcessElement
+    public void processElement(ProcessContext processContext, BoundedWindow window)
+        throws Exception {
+      String key = processContext.element();
+      bufferedKeys.add(key);
+      this.window = window;
+      this.lastMsg = processContext.timestamp();
+      if (bufferedKeys.size() > getBatchSize()) {
+        List<KV<String, String>> kvs = fetchAndFlush();
+        for (KV<String, String> kv : kvs) {
+          processContext.output(kv);
+        }
+      }
+    }
+
+    private List<KV<String, String>> fetchAndFlush() {
+      String[] keys = new String[bufferedKeys.size()];
+      bufferedKeys.toArray(keys);
+      List<String> results = jedis.mget(keys);
+      assert bufferedKeys.size() == results.size();
+      List<KV<String, String>> kvs = new ArrayList<>(bufferedKeys.size());
+      for (int i = 0; i < bufferedKeys.size(); i++) {
+        if (results.get(i) != null) {
+          kvs.add(KV.of(bufferedKeys.get(i), results.get(i)));
+        }
+      }
+      bufferedKeys = new ArrayList<>();
+      return kvs;
+    }
+
+    @FinishBundle
+    public void finishBundle(FinishBundleContext context) throws Exception {
+      List<KV<String, String>> kvs = fetchAndFlush();
+      for (KV<String, String> kv : kvs) {
+        context.output(kv, lastMsg, window);
```

Comment:
@iemejia Not sure about this since I am using the Instant and window from the last processed message to produce output in the finish bundle method.

Reply:
Though here, messages with the same window are being bundled together, stored and finally processed in the finishBundle.
In the case of ReadFn, the idea was to buffer requests till the batch size and process them at that time. Hence the output is pushed in both processElement method as well as finishBundle.
Not sure if I understand how to use a map with window as the key.

Reply:
@iemejia could you please advice on the above ?